Enable arm64 builds for Linux

praveenkumar opened the pull request.

praveenkumar committed:
packaging: Don't build rpm for ppc64le arch

ppc64le architecture not supported by OpenShift Local as of now and it
also create confusion because ppc64le package contains the amd64 binary.

praveenkumar committed:
Add support to generate aarch64 bits for Linux

This pr adds linux arm64 target to makefile and also make change to spec
file to generate right package for arm64.

praveenkumar commented:
/hold

cfergeau commented on packaging/rpm/crc.spec.in:
```diff
@@ -24,7 +24,7 @@ Name:           %{goname}
 Summary:        CRC's main executable
 License:        APL 2.0
-ExcludeArch:    armv7hl i686
+ExcludeArch:    armv7hl i686 aarch64 ppc64le
```
I'd move this after the next commit..

praveenkumar replied:
you mean first commit should be enable arm bits and then this one?

cfergeau replied:
Yes, it's odd for me to first remove `aarch64`, and then readd it in the commit right after

praveenkumar replied:
Done.

cfergeau commented on packaging/rpm/crc.spec.in:
```diff
@@ -65,7 +65,12 @@ make COMMIT_SHA=__COMMIT_SHA__ GO_EXTRA_LDFLAGS="-B 0x$(head -c20 /dev/urandom|o
 install -m 0755 -vp %{gobuilddir}/src/%{goipath}/out/linux-amd64/crc %{buildroot}%{_bindir}/
+%endif
+%ifarch aarch64
+install -m 0755 -vp %{gobuilddir}/src/%{goipath}/out/linux-arm64/crc %{buildroot}%{_bindir}/
```
Could this use `go env GOARCH` instead of hardcoding these `ifarch`?

praveenkumar replied:
didn't find how to run that in the `install` section? https://community.nethserver.org/t/cross-compiling-multi-arch-rpms/18493 is also suggest same what we are using.

praveenkumar replied:
Do you think https://github.com/kovyrin/rpmbuild/blob/master/golang/SPECS/golang.spec#L27-L38 is better what we are having now?

cfergeau replied:
> Do you think https://github.com/kovyrin/rpmbuild/blob/master/golang/SPECS/golang.spec#L27-L38 is better what we are having now?

Yes I'd prefer something like this, it's a lot easier to see what is arch-specific, and what is common.

gbraad commented:
Were other related issues addressed, like arm versions of the helper and driver?

praveenkumar commented:
> Were other related issues addressed, like arm versions of the helper and driver?

not yet, I will look into this again today, was destructed by 4.15 bundle creation and other issues.

cfergeau commented:
> was destructed by 4.15 bundle creation

Hopefully you were only distracted ;) ( https://en.wiktionary.org/wiki/destruct )

praveenkumar committed:
constants/constants.go: Add arch info to admin-helper linux binary

admin-helper now have arm64 and amd64 binary for linux platform. This PR
will make sure correct binary is downloaded and used for specific
platform.

praveenkumar committed:
libvirt/constants.go: Add arch info to download machine-driver-libvirt binary

machine-driver-libvirt now have arm64 and amd64 binary for linux platfrom. This PR
will make sure correct binary is downloaded and used for specific
platform. Name of the binary should be `machine-driver-libvirt` because
this name is hard-coded as part of the plugin

- https://github.com/crc-org/machine/blob/master/libmachine/drivers/plugin/localbinary/plugin.go#L55-L63

Now during start user get follow warning which need to be ignored
```
WARN Executable name is crc-driver-libvirt but extracted file name is crc-driver-libvirt-amd64
```

gbraad commented:
What is the status of this overall? Seems this took a lot more effort than was anticipated.

gbraad commented on Makefile:
```diff
@@ -106,14 +106,17 @@ $(BUILD_DIR)/macos-arm64/crc: $(SOURCES)
 
 .PHONY: cross ## Cross compiles all binaries
-cross: $(BUILD_DIR)/macos-arm64/crc $(BUILD_DIR)/macos-amd64/crc $(BUILD_DIR)/linux-amd64/crc $(BUILD_DIR)/windows-amd64/crc.exe
+cross: $(BUILD_DIR)/macos-arm64/crc $(BUILD_DIR)/macos-amd64/crc $(BUILD_DIR)/linux-amd64/crc $(BUILD_DIR)/linux-arm64/crc $(BUILD_DIR)/windows-amd64/crc.exe
```
Not ready for this. Please leave this as a specific target.

praveenkumar replied:
Done, just have specific target not as part of cross.

praveenkumar committed:
Makefile: Use GOARCH for linux-release target

This is the target which is used as part of `release` and currently
hardcoded to `amd64` which means even this target is run on `arm64`
machine it will still create the `amd64` binary. In spec file also
`release` target is used and for `arm64` builds `amd64` binary is used
without this fix.

cfergeau commented on pkg/crc/preflight/preflight_checks_unix.go:
```diff
@@ -139,8 +139,8 @@ func fixAdminHelperExecutableCached() error {
-	// Only supported arches are amd64, and arm64 on macOS
-	if runtime.GOARCH == "amd64" || (runtime.GOARCH == "arm64" && runtime.GOOS == "darwin") {
+	// Only supported arches are amd64, and arm64 on macOS & Linux
+	if runtime.GOARCH == "amd64" || (runtime.GOARCH == "arm64" && runtime.GOOS != "windows") {
```
Might be easier to read as a switch case
```
switch runtime.GOARCH {
  case "amd64":
    return nil
  case "arm64":
    if runtime.GOOS != "windows" {
        return nil
    }
}
```

I'd keep explicit linux/darwin checks instead of `runtime.GOOS != "windows"`  to avoid returning `nil` on freebsd if someone tries that. This also alllow to add a warning for linux/arm64 that it's not widely tested.

praveenkumar replied:
Done and add the warning for linux/arm64

cfergeau commented on pkg/crc/preflight/preflight_checks_unix.go:
```diff
@@ -139,8 +139,8 @@ func fixAdminHelperExecutableCached() error {
+	if runtime.GOARCH == "amd64" || (runtime.GOARCH == "arm64" && runtime.GOOS != "windows") {
 		return nil
 	}
 	return fmt.Errorf("CRC can only run on AMD64/Intel64 CPUs and Apple silicon")
```
This needs to be updated.

praveenkumar replied:
This should be updated once we have support for arm64 on linux, as of now warning should be OK?

cfergeau replied:
Yes, with the warning this works.

cfergeau commented on pkg/crc/constants/constants.go:
```diff
@@ -55,7 +55,7 @@ const (
 var adminHelperExecutableForOs = map[string]string{
 	"darwin":  "crc-admin-helper-darwin",
-	"linux":   "crc-admin-helper-linux",
+	"linux":   fmt.Sprintf("crc-admin-helper-linux-%s", runtime.GOARCH),
```
There's a "platfrom" typo in the commit log.
I don't think the old binary name is cleaned up on upgrades?

praveenkumar replied:
> I don't think the old binary name is cleaned up on upgrades?

No, older binary is not cleaned up as part of upgrade. Should we do it as part of preflight checks or as part of cache.go ?

praveenkumar replied:
> There's a "platfrom" typo in the commit log.

Done.

cfergeau replied:
> Should we do it as part of preflight checks or as part of cache.go ?

In the past we've done this as part of preflight checks.

cfergeau commented on packaging/rpm/crc.spec.in:
```diff
@@ -30,7 +30,7 @@ Name:           %{goname}
 Summary:        CRC's main executable
 License:        APL 2.0
-ExcludeArch:    armv7hl i686
+ExcludeArch:    armv7hl i686 ppc64le
```
> it also create confusion because ppc64le package contains the amd64 binary.

I'd prefer to fix this rather than add artificial hurdles to ppc64le enablement if someone ever wants to try this. 

If you add this, you should also remove
```
# crc uses `go test -race`, which triggers gvisor issues on ppc64le:
# vendor/gvisor.dev/gvisor/pkg/sync/race_unsafe.go:47:6: missing function body
# and a ppc64le implementation is indeed missing on ppc64le:
# https://github.com/google/gvisor/blob/master/pkg/sync/race_unsafe.go
# https://github.com/google/gvisor/tree/master/pkg/sync
%ifnarch ppc64le
make test
%endif
```

praveenkumar replied:
Removed it and also remove this `ppc64le` part for not running the `make test`.

cfergeau commented on packaging/rpm/crc.spec.in:
```diff
@@ -30,7 +30,7 @@ Name:           %{goname}
-ExcludeArch:    armv7hl i686
+ExcludeArch:    armv7hl i686 ppc64le
 URL:            %{gourl}
 ExcludeArch:    s390x
```
Why is this separate?

praveenkumar replied:
No idea, was there before but I have now removed one of that.

cfergeau commented on Makefile:
```diff
@@ -310,20 +310,20 @@ release: clean linux-release macos-release-binary windows-release-binary check
+	$(HOST_BUILD_DIR)/crc-embedder embed --log-level debug --goos=linux $(BUILD_DIR)/linux-${GOARCH}/crc
 else
-	$(HOST_BUILD_DIR)/crc-embedder embed --log-level debug --cache-dir=$(EMBED_DOWNLOAD_DIR) --no-download --goos=linux $(BUILD_DIR)/linux-amd64/crc
+	$(HOST_BUILD_DIR)/crc-embedder embed --log-level debug --cache-dir=$(EMBED_DOWNLOAD_DIR) --no-download --goos=linux $(BUILD_DIR)/linux-${GOARCH}/crc
```
Have you checked that after the naming changes for crc-admin-helper and crc-driver-libvirt, the embedding process still works fine, and that `crc setup` is still able to find/use the embedded files (and not download them) if they are not cached locally?
It looks like this should work, but I may have missed something.

praveenkumar replied:
I will check and let you know.

praveenkumar replied:
Tested and it works.
```
$ make embed_crc_helpers 
go build --tags="build" -ldflags="-X github.com/crc-org/crc/v2/pkg/crc/version.crcVersion=2.37.0 -X github.com/crc-org/crc/v2/pkg/crc/version.ocpVersion=4.15.17 -X github.com/crc-org/crc/v2/pkg/crc/version.okdVersion=4.15.0-0.okd-2024-02-23-163410 -X github.com/crc-org/crc/v2/pkg/crc/version.microshiftVersion=4.15.17 -X github.com/crc-org/crc/v2/pkg/crc/version.commitSha=b0972c " -o out/linux-amd64/crc-embedder  ./cmd/crc-embedder
out/linux-amd64/crc-embedder embed --log-level debug --goos=linux out/linux-amd64/crc
DEBU Downloading https://github.com/crc-org/machine-driver-libvirt/releases/download/0.13.8/crc-driver-libvirt-amd64 to /tmp/crc-embedder647635193 
DEBU Download saved to /tmp/crc-embedder647635193/crc-driver-libvirt-amd64 
DEBU Downloading https://github.com/crc-org/admin-helper/releases/download/v0.5.4/crc-admin-helper-linux-amd64 to /tmp/crc-embedder647635193 
DEBU Download saved to /tmp/crc-embedder647635193/crc-admin-helper-linux-amd64 
DEBU Embedding /tmp/crc-embedder647635193/crc-driver-libvirt-amd64 in out/linux-amd64/crc 
DEBU Embedding /tmp/crc-embedder647635193/crc-admin-helper-linux-amd64 in out/linux-amd64/crc 
```

and then
```
$ ./crc setup --log-level debug
DEBU CRC version: 2.37.0+b0972c                   
DEBU OpenShift version: 4.15.17                   
DEBU Running 'crc setup'                          
INFO Using bundle path /home/prkumar/.crc/cache/crc_libvirt_4.15.17_amd64.crcbundle 
DEBU resolv.conf file path: /etc/resolv.conf      
INFO Checking if running as non-root              
INFO Checking if running inside WSL2              
INFO Checking if crc-admin-helper executable is cached 
DEBU crc-admin-helper executable is not cached    
INFO Caching crc-admin-helper executable          
DEBU Trying to extract crc-admin-helper-linux-amd64 from crc executable 
DEBU Extracting embedded 'crc-admin-helper-linux-amd64' from /home/prkumar/work/github/crc/out/linux-amd64/crc to /tmp/crc342705720/crc-admin-helper-linux-amd64 
DEBU Copying '/tmp/crc342705720/crc-admin-helper-linux-amd64' to '/home/prkumar/.crc/bin/crc-admin-helper-linux-amd64' 
DEBU crc-admin-helper executable cached           
DEBU Making /home/prkumar/.crc/bin/crc-admin-helper-linux-amd64 suid 
INFO Using root access: Changing ownership of /home/prkumar/.crc/bin/crc-admin-helper-linux-amd64 
DEBU Running '/usr/bin/sudo chown root /home/prkumar/.crc/bin/crc-admin-helper-linux-amd64' 

```

praveenkumar commented:
/unhold
